jax/experimental/jax2tf/saved_model_lib.py

```diff
@@ -101,17 +87,17 @@ def jax_fn_for_save_model(params, inputs):
       raise ValueError("For shape-polymorphic conversion a single "
                        "input_signature is supported.")
   tf_fn = jax2tf.convert(
-      jax_fn,
-      with_gradient=with_gradient,
-      in_shapes=[None, shape_polymorphic_input_spec],
-      enable_xla=enable_xla)
+    jax_fn,
+    with_gradient=with_gradient,
+    in_shapes=[None, shape_polymorphic_input_spec],
+    enable_xla=enable_xla)
 
   # Create tf.Variables for the parameters.
   param_vars = tf.nest.map_structure(
-      # If with_gradient=False, we mark the variables behind as non-trainable,
-      # to ensure that users of the SavedModel will not try to fine tune them.
-      lambda param: tf.Variable(param, trainable=with_gradient),
-      params)
+    # If with_gradient=False, we mark the variables behind as non-trainable,
```

Comment:
This seems odd to me, if the model is not trainable wouldn't it be better to keep the params as tensors baked into the graph (e.g. `if not with_gradient: param_vars = params`)? The typical use for non-trainable variables is for things like moving averages (not for model parameters) so I am a bit surprised to see them used in this way.

Using non-trainable tensors here may also impact how models that use non-trainable variables (e.g. batch norm) need to be handled. Right now this API only supports closing over batch norm state (which I believe will be typically fine for inference).

Reply:
I thought that even for non-trainable variables it is better to keep them separate from the graph as a matter of SavedModel organization. Perhaps it would only make a difference when the size of the variables are larger than the proto size. 

If the user wants to keep the parameters baked into the graph, they can do it before calling this function: pass [] for params. 

BTW, I have not looking into how this would be used with BatchNorm.

Reply:
That makes sense to me, I would still suggest `params` are represented as `trainable` (even if we disallow gradients) just to allow us to represent non-trainable mutable values in the future.

The `trainable` bit in TensorFlow is a bit of a weird beast, for example in Keras if you mark a `Layer` as non-trainable, it's parameters still have `trainable=True` (since this bit is immutable and set at construction time). Personally I think of this bit as `parameter=True`.

```python
import tensorflow as tf
d = tf.keras.layers.Dense(10)
d(tf.ones([1, 1]))

assert len(d.trainable_variables) == 2
assert len(d.variables) == 2

d.trainable = False  # Layer.trainable != Variable.trainable
assert len(d.trainable_variables) == 0
assert len(d.variables) == 2
assert all(v.trainable for v in d.variables)
```

Reply:
It turns out that Keras insists on wanting to take gradients through these variables even if the layer is marked as non-trainable.